### PR TITLE
[WIP] Integrate Dynamo + SPMD for Inference

### PR DESCRIPTION
**Pull request by steventk-g**
Includes @yeounoh's changes to use SPMD with `torch.compile`

### CODE REVIEW

File: test/spmd/test_dynamo_spmd_inference_latency.py [hunk 1]
```diff
@@ -0,0 +1,114 @@
+"""
```
Comment (yeounoh) [inline test/spmd/test_dynamo_spmd_inference_latency.py:1]: Let's remove this script before merging

File: test/spmd/test_inference_spmd_dynamo_imagenet.py [hunk 2]
```diff
@@ -0,0 +1,237 @@
...
+
+
+def inference_imagenet():
+  print('==> Preparing data..')
+  img_dim = get_model_property('img_dim')
+  device = xm.xla_device()
+  if FLAGS.fake_data:
+    assert FLAGS.test_set_batch_size == 1
+    test_loader = xu.SampleGenerator(
+        data=(torch.zeros(FLAGS.test_set_batch_size, 3, img_dim, img_dim).to(device),
```
Comment (yeounoh) [inline test/spmd/test_inference_spmd_dynamo_imagenet.py:96]: Maybe we comment on that we needed `to(device)` here for spmd?

**Review by yeounoh — CHANGES_REQUESTED** [review]
Mostly LGTM, added a few minor comments. Also we need to address linter checks.

File: test/spmd/test_inference_spmd_dynamo_imagenet.py [hunk 3]
```diff
@@ -0,0 +1,238 @@
...
+  model = torchvision.models.resnet50().to(
+      device)  # get_model_property('model_fn')().to(device)
+
+  input_mesh = None
+  if FLAGS.sharding:
+    num_devices = pjrt.global_device_count()
+    device_ids = np.arange(num_devices)
+    # Model sharding
+    if 'conv' in FLAGS.sharding:
+      # Shard the model's convlution layers along two dimensions
```
Comment (yeounoh) [inline test/spmd/test_inference_spmd_dynamo_imagenet.py:136]: typo "convolution"

File: test/spmd/test_inference_spmd_dynamo_imagenet.py [hunk 4]
```diff
@@ -0,0 +1,238 @@
...
+        test_dataset,
+        batch_size=FLAGS.test_set_batch_size,
+        sampler=None,
+        drop_last=FLAGS.drop_last,
+        shuffle=False)
+
+  torch.manual_seed(42)
+
+  model = torchvision.models.resnet50().to(
+      device)  # get_model_property('model_fn')().to(device)
```
Comment (yeounoh) [inline test/spmd/test_inference_spmd_dynamo_imagenet.py:128]: Need to clean this up, either use `get_model_property('model_fn')().to(device)` or stick to `resnet50()`, in which case we don't need some of the model props code above.

File: test/dynamo/test_bridge.py [hunk 5]
```diff
@@ -9,6 +9,7 @@
 
 import torch_xla.core.dynamo_bridge as bridge
 import torch_xla.core.xla_model as xm
+import torch_xla.experimental.xla_sharding as xs
```
Comment (yeounoh) [inline test/dynamo/test_bridge.py:12]: this is needed?

File: test/dynamo/test_bridge.py [hunk 6]
```diff
@@ -108,6 +109,7 @@ def test_wrapper(self):
     optimized_mod = bridge.extract_compiled_graph(
         fx.symbolic_trace(xla_module), inputs)
 
+    print('\nCompiled graph extraction done!\n')
```
Comment (yeounoh) [inline test/dynamo/test_bridge.py:112]: this is a debugging probe?

File: test/tpu/xla_test_job.yaml [hunk 7]
```diff
@@ -44,7 +44,8 @@ spec:
       python3 pytorch/xla/test/pjrt/test_experimental_pjrt_tpu.py
       python3 pytorch/xla/test/spmd/test_xla_sharding.py
       python3 pytorch/xla/test/spmd/test_xla_virtual_device.py
-      python3 pytorch/xla/test/spmd/test_train_spmd_linear_model.py
+      XLA_USE_SPMD=1 python3 pytorch/xla/test/spmd/test_train_spmd_linear_model.py --sharding batch
+      XLA_USE_SPMD=1 python3 pytorch/xla/test/spmd/test_inference_spmd_dynamo_imagenet.py  --fake_data --use_dynamo --sharding batch
```
Comment (yeounoh) [inline test/tpu/xla_test_job.yaml:48]: awesome, thank you!

File: torch_xla/core/dynamo_bridge.py [hunk 8]
```diff
@@ -190,6 +190,10 @@ def add_nones(self, value_list):
 
 
 def is_xla_tensor(tensor: torch.Tensor) -> bool:
+  # TODO(yeounoh) check if tensor sharding annotation can be accessed here
```
Comment (yeounoh) [inline torch_xla/core/dynamo_bridge.py:193]: I think we can remove my comment here now.

**Review by yeounoh — CHANGES_REQUESTED** [review]
Added more comments

File: test/spmd/test_inference_spmd_dynamo_imagenet.py [hunk 9]
```diff
@@ -0,0 +1,238 @@
...
+        )
+      test_loader = pl.MpDeviceLoader(
+          test_loader,
+          device,
+          input_sharding=xs.ShardingSpec(input_mesh, (0, 1, 2, 3)))
+
+  model.eval()
+  xm.mark_step()
+  xm.wait_device_ops()
+  met.clear_all()
```
Comment (JackCaoG) [inline test/spmd/test_inference_spmd_dynamo_imagenet.py:191]: why are these needed? I had these in my dynamo test because test data is fake so I need a `mark_step` to materalize the fake input, if that;'s not the case here we don't need to do a mark_step here.

File: test/spmd/test_inference_spmd_dynamo_imagenet.py [hunk 10]
```diff
@@ -0,0 +1,238 @@
...
+    writer = test_utils.get_summary_writer(FLAGS.logdir)
+
+  def inference_loop_fn(loader):
+    start_0 = time.time()
+    for step, (data, _) in enumerate(loader):
+      if step == 1:
+        start_1 = time.time()
+      output = model(data)
+    end = time.time()
+    return end - start_0, end - start_1
```
Comment (JackCaoG) [inline test/spmd/test_inference_spmd_dynamo_imagenet.py:208]: If this test is inference only, we should surface that in the test name.

File: test/spmd/test_inference_spmd_dynamo_imagenet.py [hunk 11]
```diff
@@ -0,0 +1,238 @@
+import args_parse
```
Comment (JackCaoG) [inline test/spmd/test_inference_spmd_dynamo_imagenet.py:1]: We also need to basic test similar to https://github.com/pytorch/xla/blob/master/test/dynamo/test_dynamo.py#L45 to check the correctness of the dynamo + SPMD. In this case I guess you should expect `dynamo + SPMD` yield the same result as `SPMD` along to test this integration. Test can be either under `SPMD` dir or `dynamo` dir. Currently we don't even know if SPMD is emitting the correct result when used with dynamo(I suppose we did some e2e test).

File: test/tpu/xla_test_job.yaml (hunk 7 above)
Comment (JackCaoG) [reply test/tpu/xla_test_job.yaml:48]: How long will this test take? It is inference so I guess it won't be too bad but I am trying to prevent us from adding too many burden to the already very long CI.

File: torch_xla/csrc/xla_graph_executor.cpp [hunk 12]
```diff
@@ -613,17 +621,41 @@ XLAGraphExecutor::ExecuteComputationWithBarrier(
   std::shared_ptr<XLAGraphExecutor::Async> async = std::make_shared<Async>(
       &coll, std::move(arguments), placeholders, std::move(cachedComputation));
 
-  auto syncfn = [async, hash]() {
+  auto syncfn = [async, hash, sharding_specs]() {
```
Comment (JackCaoG) [inline torch_xla/csrc/xla_graph_executor.cpp:624]: why `sharding_specs` needs to be passed from original function? It is init as default value right? We might as well just init it within this function.

Comment (JackCaoG) [reply torch_xla/csrc/xla_graph_executor.cpp:624]: ok, I guess the point here is we need to supply a proper sharding spec for the output, but it is currently not supported yet. As a follow up, what's the right way to determine the sharding spec for output?

**Review by JackCaoG — COMMENTED** [review]
mostly LGTM.

FYI @wonjoolee95 in this pr we break a really important assumption of the dynamo which is "same fx graph will result in same IR/HLO graph". In the SPMD case, even if the fx graph is the same, we can annotate the input XLATensor and make it generate different IR/HLO graph.

There is one thing that is still unclear to me and I hope @yeounoh and maybe @alanwaketan can give me some answer. AFAIK, during the inital phase of the dynamo `extract_compile_graph`, dynamo will pass us the fx graph along with the fake XLATensor with same shape with the real model. Does dyanmo somehow correctly move the sharding annotation from the real XLATensor to the Fake XLATensor? Otherwise I would expect we hit a cache miss when we execute the real SPMD computation.

File: test/spmd/test_inference_spmd_dynamo_imagenet.py (hunk 3 above)
Comment (steventk-g) [reply test/spmd/test_inference_spmd_dynamo_imagenet.py]: Done

File: test/dynamo/test_bridge.py (hunk 5 above)
Comment (steventk-g) [reply test/dynamo/test_bridge.py]: Removed

File: test/dynamo/test_bridge.py (hunk 6 above)
Comment (steventk-g) [reply test/dynamo/test_bridge.py]: Removed

File: test/dynamo/test_bridge.py [hunk 13]
```diff
@@ -117,6 +119,8 @@ def test_wrapper(self):
       # make sure above lazy computation is executed.
       xm.mark_step()
 
+      print(f'\nexpected: {expected}\n')
```
[the remark this answers is not in the document]
Comment (steventk-g) [reply test/dynamo/test_bridge.py]: Removed

File: test/spmd/test_inference_spmd_dynamo_imagenet.py (hunk 10 above)
Comment (steventk-g) [reply test/spmd/test_inference_spmd_dynamo_imagenet.py:219]: Where should we add a reference to inference? The main function in this file is `inference_imagenet` and the file name is `test/spmd/test_inference_spmd_dynamo_imagenet.py`

File: test/spmd/test_inference_spmd_dynamo_imagenet.py (hunk 9 above)
Comment (steventk-g) [reply test/spmd/test_inference_spmd_dynamo_imagenet.py:202]: We need to materialize the input here as well. @yeounoh may be able to add more context

Comment (JackCaoG) [reply test/spmd/test_inference_spmd_dynamo_imagenet.py:202]: ok, maybe add a comment above to mentioned that this `mark_step` is to materialize the fake input then.

File: test/spmd/test_inference_spmd_dynamo_imagenet.py (hunk 10 above)
Comment (JackCaoG) [reply test/spmd/test_inference_spmd_dynamo_imagenet.py:219]: oh I miss the inference in the name, my bad.

**Comment by JackCaoG** [comment]
@steventk-g let me know when you want me to take another look

File: test/spmd/test_inference_spmd_dynamo_imagenet.py (hunk 11 above)
Comment (steventk-g) [reply test/spmd/test_inference_spmd_dynamo_imagenet.py:1]: Tried adding a test, currently crashing on `res_xla_spmd_dynamo = self.fn_simple_dynamo(sharded_x, sharded_y)`

File: test/spmd/test_inference_spmd_dynamo_imagenet.py (hunk 9 above)
Comment (steventk-g) [reply test/spmd/test_inference_spmd_dynamo_imagenet.py:202]: Done

File: torch_xla/csrc/xla_graph_executor.cpp (hunk 12 above)
Comment (steventk-g) [reply torch_xla/csrc/xla_graph_executor.cpp:623]: Good question, I'm not sure the right way to determine sharding spec for output. @yeounoh might be able to answer or provide more context

Comment (JackCaoG) [reply torch_xla/csrc/xla_graph_executor.cpp:623]: That's fine, we will leave that as a follow up.

File: test/spmd/test_inference_spmd_dynamo_imagenet.py (hunk 11 above)
Comment (JackCaoG) [reply test/spmd/test_inference_spmd_dynamo_imagenet.py:1]: let me know if you need help.

**Comment by JackCaoG** [comment]
I am taking over this commit, since it is on @steventk-g 's fork and it already has conflcits with the master, I will just open a new pr and patch from master.

**Comment by alanwaketan** [comment]
> I am taking over this commit, since it is on @steventk-g 's fork and it already has conflcits with the master, I will just open a new pr and patch from master.

Let me know when it's ready. I can review it.

**Comment by JackCaoG** [comment]
Close this one in favor of https://github.com/pytorch/xla/pull/5002
